### DISCUSSION TITLE
Quick question on iris.load callback and constraints

### DISCUSSION THREAD

**Discussion by larsbarring** [SciTools/iris#4185] [Q&A] [answered]
If I am loading a netcdf file using both a callback function and a constraint, which is done first? 
Context: I want to extract a small region from a rather poorly metadata formatted large file where the x,y coordinates are off by a small fraction that I need to adjust. So can I use the callback to tweak the coordinated to what they are supposed to be and then let the constraints do its work to cut out the subdomain?

**Comment by trexfeathers** [comment]
Just spent 5mins looking through the Iris source and I'm pretty sure it's constraints first, then callbacks.

**Reply by larsbarring** [reply]
Big thanks for the superquick answer !

**Reply by pp-mo** [reply, marked as answer]
Actually, I just spent **_20_** minutes chasing this, and I think constraints come **_afterwards_** ...

The callback is invoked within the low-level loader (e.g. [here](https://github.com/SciTools/iris/blob/master/lib/iris/fileformats/netcdf.py#L816-L817), or [here](https://github.com/SciTools/iris/blob/master/lib/iris/fileformats/rules.py#L372-L376)), but the constraints are mostly used by the "_CubeFilterCollection.from_cubes" call in [iris._load_collection](https://github.com/SciTools/iris/blob/master/lib/iris/__init__.py#L268-L276), 
This underlies the various iris.load_xxx functions, but it works on "raw cubes" delivered by the loader.

I say "mostly", because the constraints _are_ passed down to the low-level loaders.  
But I think that is only so that "special cases" can be interpreted, to filter file data and speed up loading, [notably, here](https://github.com/SciTools/iris/blob/master/lib/iris/fileformats/pp.py#L2129-L2130), but see also recent #4176.
But those are, by definition, "exceptional" cases.

----

Whatever, this is an amazingly pertinent  question + we really should make this all this clearer somewhere !

**Reply by rcomer** [reply]
At #3944 @bjlittle said it’s constraints first 🤔

**Reply by larsbarring** [reply]
Well, I realise why my feeble attempts at reading the code were in vain. I did look at the documentation, but it never appeared to me to check the git issues... 
BTW: The new readthedocs documentation is  really a quantum leap forward!

**Reply by SimonPeatman** [reply]
> At #3944 @bjlittle said it’s constraints first

However, the [code I posted](https://github.com/SciTools/iris/issues/3944#issuecomment-784459170) there implies that's not true ...

Either way, I agree with @pp-mo that this really needs to be clarified.
